lib/ansible/modules/cloud/vmware/vmware_object_rename.py

```diff
@@ -0,0 +1,172 @@
+#
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import absolute_import, division, print_function
+__metaclass__ = type
+
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.1',
+    'status': ['preview'],
+    'supported_by': 'community'
+}
+
+
+DOCUMENTATION = '''
+---
+module: vmware_object_rename
+short_description: Rename a vmware object
+description:
+    - This module can be used to rename a vmware object.
+version_added: 2.6
+author:
+    - Abhijeet Kasurde (@akasurde) <akasurde@redhat.com>
+notes:
+    - Tested on vSphere 6.5
+requirements:
+    - "python >= 2.6"
+    - PyVmomi
+options:
+   name:
+     description:
+     - Name of the object to rename.
+     required: True
+   new_name:
+     description:
+     - New name of the object.
+     required: True
+   object_type:
+     description:
+     - Type of object to rename.
+     required: False
+     default: 'VirtualMachine'
+     choices: ['VirtualMachine', 'Datacenter', 'ClusterComputeResource',
+               'DistributedVirtualSwitch', 'HostSystem', 'Datastore',
+               'Network', 'DistributedVirtualPortgroup']
+extends_documentation_fragment: vmware.documentation
+'''
+
+EXAMPLES = '''
+- name: Rename Datacenter
+  vmware_object_rename:
+    hostname: 192.168.1.209
+    username: administrator@vsphere.local
+    password: vmware
+    validate_certs: False
+    name: datacenter1
+    new_name: datacenter2
+    object_type: 'Datacenter'
+  delegate_to: localhost
+
+- name: Rename Cluster
+  vmware_object_rename:
+    hostname: 192.168.1.209
+    username: administrator@vsphere.local
+    password: vmware
+    validate_certs: False
+    name: cluster1
+    new_name: nw_cluster2
+    object_type: 'ClusterComputeResource'
+  delegate_to: localhost
+'''
+
+RETURN = """ # """
+
+try:
+    import pyVmomi
+    from pyVmomi import vim
+except ImportError:
+    pass
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils._text import to_native
+from ansible.module_utils.vmware import PyVmomi, vmware_argument_spec, wait_for_task
+
+
+class VmObjectMgr(PyVmomi):
+    def __init__(self, module):
+        super(VmObjectMgr, self).__init__(module)
+
+        object_types = {
+            'Datacenter': vim.Datacenter,
+            'VirtualMachine': vim.VirtualMachine,
+            'ClusterComputeResource': vim.ClusterComputeResource,
+            'DistributedVirtualSwitch': vim.DistributedVirtualSwitch,
+            'HostSystem': vim.HostSystem,
+            'Datastore': vim.Datastore,
+            'Network': vim.Network,
+            'DistributedVirtualPortgroup': vim.dvs.DistributedVirtualPortgroup,
+        }
+
+        old_name = module.params['name']
+        new_name = module.params['new_name']
+        object_type = object_types[module.params['object_type']]
+
+        managed_objects = self.get_managed_objects_properties(vim_type=object_type, properties=['name'])
```

Comment:
In a large environment, you will end up with a very slow action here... I know of deployments where hosts have close to 500 DVS portgroups for instance. 

Not sure if in some cases, SearchIndex.findChild could help (i don't exactly know for which objects this works, it seems typically used for VMs and folders, but it really does not limit that according to the docs, seems like any entity type can work)